lib/src/user_config.rs

```diff
@@ -0,0 +1,243 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A wrapper around retrieval and storage of secure user configuration.
+
+use std::fmt::Debug;
+use std::io::Write as _;
+use std::path::Path;
+
+use ed25519_dalek::Signer as _;
+use ed25519_dalek::SigningKey;
+use etcetera::BaseStrategy as _;
+use pkcs8::DecodePrivateKey as _;
+use pkcs8::EncodePrivateKey as _;
+use pkcs8::EncodePublicKey as _;
+use prost::Message as _;
+use rand::RngCore as _;
+use rand::TryRngCore as _;
+use rand::rngs::OsRng;
+use tempfile::NamedTempFile;
+use thiserror::Error;
+
+use crate::file_util::IoResultExt as _;
+use crate::file_util::PathError;
+use crate::protos::user_config::RepoConfig;
+use crate::protos::user_config::SecureUserConfig;
+use crate::protos::user_config::WorkspaceConfig;
+
+#[derive(Error, Debug)]
+/// Error occurred while dealing with user configs
+pub enum UserConfigError {
+    /// Failed to determine the home directory
+    #[error("Unable to determine the home directory.")]
+    HomeDirNotFound,
+    /// Failed to read / write to the specified path
+    #[error(transparent)]
+    PathError(#[from] PathError),
+    /// Failed to load the private key.
+    #[error(transparent)]
+    PrivateKeyError(#[from] ed25519_dalek::pkcs8::Error),
+    /// Failed to load the public key.
+    #[error(transparent)]
+    PublicKeyError(#[from] ed25519_dalek::pkcs8::spki::Error),
+
+    /// Failed to decoode the user configuration file.
+    #[error(transparent)]
+    DecodeError(#[from] prost::DecodeError),
+
+    /// Missing the user configuration file.
+    #[error(
+        "This repo was generated by an old version of jj and does not have secure user \
+         configuration"
+    )]
+    LegacyRepo,
+}
+
+/// A config file represented by a string instead of backed by a file.
+pub trait ConfigType: prost::Message + Default + Clone + Debug {
+    /// The name of this type.
+    fn kind() -> &'static str;
+    /// The filename of the configuration file to write.
+    fn filename() -> &'static str;
+    /// The string containing the config.
+    fn config(&self) -> &str;
+}
+
+impl ConfigType for RepoConfig {
+    fn kind() -> &'static str {
+        "repo"
+    }
+
+    fn filename() -> &'static str {
+        "config.binarypb"
+    }
+
+    fn config(&self) -> &str {
+        &self.config
+    }
+}
+
+impl ConfigType for WorkspaceConfig {
+    fn kind() -> &'static str {
+        "workspace"
+    }
+
+    fn filename() -> &'static str {
+        "workspace-config.binarypb"
+    }
+
```

Comment:
nit: these should end with `binpb` per Google's (resp: the protobuf teams) own suggestion in https://protobuf.dev/programming-guides/techniques/